### DISCUSSION TITLE
Modify Asset with a Runtime

### DISCUSSION THREAD

**Discussion by 101arrowz** [parcel-bundler/parcel#4656] [General] [answered]
Is it possible to add to an existing asset (e.g. appending HTML tags to the `<head>` or adding some JavaScript to be run before anything else) with a runtime?

**Comment by mischnic** [comment, marked as answer]
Runtimes can add additional assets, so you can add a javascript asset that runs before a bundle's actual entrypoint is run:

https://github.com/parcel-bundler/parcel/blob/d33fcd9cf826188288e7ffb5dfecbfa221592458/packages/runtimes/react-refresh/src/ReactRefreshRuntime.js#L31-L35

In the case of HTML, the whole file is one asset (since 1 HTML source file becomes 1 HTML bundle), and multiple HTML assets in the same bundle don't have a defined order. (Note that this is a rather arbitrary implementation decision in `@parcel/packager-html`, you could write your own packager.)

WIP (plugin) docs: https://parcel2-docs.now.sh/plugin-system/

**Reply by 101arrowz** [reply]
You're suggesting I override the HTML packager if I want to insert custom HTML. Is it possible to have both a packager and a runtime in the same package though? I'm trying to port [this package](https://github.com/101arrowz/pwa-manifest/tree/master/packages/parcel-plugin-pwa-manifest) to Parcel v2, and it needs to both add new assets and modify the HTML files, whereas Parcel v2 only seems to allow one type of plugin.

**Reply by mischnic** [reply]
No, this is what config "plugins" are for:
```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{js,mjs,jsm,jsx,es6,cjs,ts,tsx}": [
      "@prefresh/parcel-transformer-preact-refresh-babel",
      "...",
      "@prefresh/parcel-transformer-preact-refresh-wrap"
    ]
  },
  "runtimes": {
    "browser": [
      "...", "@prefresh/parcel-runtime-preact-refresh"
    ]
  }
}
```
(https://github.com/mischnic/preact-refresh/tree/4e1f8be8a20014f40d55ad091660e4e5e8e36bd1/packages/parcel-config)

https://parcel2-docs.now.sh/plugin-system/config/

**Reply by 101arrowz** [reply]
I just realized that transformer plugins can do exactly what I need: modifying assets AND adding new ones in. So I'll use that instead of creating a whole new config package.

**Reply by 101arrowz** [reply]
@mischnic This is probably better suited in another discussion, but I'll ask here to reduce clutter: can transfomers output standalone assets like runtimes can? It seems that only child assets can be generated. The issue is that if I have two separate plugins, it's impossible to share state between them (in my case, the HTML to inject and the files to insert are generated in the same function call.)

At this point, I'm tempted just to write directly to the filesystem because the plugin system is way more convoluted and restrictive than in v1. I personally think reverting to the previous system of autoloading all `parcel-plugin-*`s and allowing those plugins access to all of the plugin APIs would make them much more useful.

**Reply by mischnic** [reply]
> The issue is that if I have two separate plugins, it's impossible to share state between them (in my case, the HTML to inject and the files to insert are generated in the same function call.)

Could you be more specific? The HTML transformer does something similar with inline assets.

We could also extend the HTML packager to somehow accept additional headers to insert (open a feature request or RFC for that), if this is what you want to achieve.

> I personally think reverting to the previous system of autoloading all parcel-plugin-*s and allowing those plugins access to all of the plugin APIs would make them much more useful.

The problem with that was that it was impossible for multiple plugins to override the default JS/HTML/... asset handling.

**Reply by 101arrowz** [reply]
Thanks for the quick reply. A more detailed explanation of my problem:

I have this function `generate()` from an external library that returns HTML to inject and files to add (this is both the filename and the file contents). It's an expensive operation, so I need to call it just once and get the result. However, no single plugin can satisfy this. Runtimes cannot modify assets, and Transformers cannot generate standalone assets (the API only allows child assets, which have names that are dependent upon the parent asset).

**Reply by mischnic** [reply]
> Transformers cannot generate standalone assets

You can also return a completely different asset in a transformer. The reason that there are usually child assets is imports to an asset need to obviously resolve to one single asset.
What do you mean "standalone asset"?

**Reply by 101arrowz** [reply]
From my testing and analysis of the codebase, when [TransformerResult](https://parcel2-docs.now.sh/plugin-system/transformer/#TransformerResult) objects are returned from transformers, they are [automatically converted to child assets no matter what](https://github.com/parcel-bundler/parcel/blob/v2/packages/core/core/src/Transformation.js#L276-L290). In fact, the filename of the asset is even overwritten by the filename of the original asset + the extension of the new asset (and a fingerprint). I need the filename to NOT be overwritten and the contents to not change.

Essentially, I have to write to the filesystem but want to do it through a parcel asset for caching purposes and prettier output. Is that possible?

**Reply by 101arrowz** [reply]
Also, second issue, Transfomers seem to only be able to return three assets (including the original asset), and any extra assets beyond 3 don't get emitted to the final output.

**Reply by mischnic** [reply]
I haven't look into your previous comment further, but

> Also, second issue, Transfomers seem to only be able to return three assets (including the original asset), and any extra assets beyond 3 don't get emitted to the final output.

Is it possible that you are returning TransformerResults that have the same type and not a unique `uniqueKey`?

https://github.com/parcel-bundler/parcel/blob/d33fcd9cf826188288e7ffb5dfecbfa221592458/packages/transformers/html/src/inline.js#L89-L99

**Reply by 101arrowz** [reply]
As it turns out, yes, it was the lack of `uniqueKey`. I assumed it would be autogenerated. Thanks again for your help! Still, it would be nice if Parcel could keep the original filename I provide rather than overwriting it with part of the parent asset.
